File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.cs
public class Solution {
    public void _EventualSafeNodes(int[][] graph, 
                                   ref List<int> vis,
                                   ref List<int> su,
                                   int startNode){
        Stack<int> stk = new Stack<int>();
        stk.Push(startNode);

        while(stk.Count() > 0){
            int currNode = stk.Peek();

            vis[currNode] = 1;
            bool allSafe = true;

            foreach(var neighbour in graph[currNode]){
                
                if(vis[neighbour] == 0){
                    stk.Push(neighbour);
                    allSafe = false;
                } else if(su[neighbour] == 0 || su[neighbour] == -1){
                    allSafe = false;
                    while(stk.Count() > 0 && stk.Peek() != currNode){
                        stk.Pop();
                    }
                    su[currNode] = 0;
                    stk.Pop();
                    break;
                }
            }
            
            if(allSafe){
                su[currNode] = 1;
                stk.Pop();
            }
        }

    }
    public IList<int> EventualSafeNodes(int[][] graph) {
        List<int> vis = new List<int>();
        vis = Enumerable.Repeat(0, graph.Length).ToList();
        List<int> su = new List<int>();
        su = Enumerable.Repeat(-1, graph.Length).ToList();

        int n = graph.Length-1;

        for(int j = n; j >= 0; j--)
        {
            if(vis[j] == 1) continue;
            _EventualSafeNodes(graph, ref vis, ref su, j);
        }

        List<int> ans = new List<int>();
        for(int i=0;i<su.Count();i++){
            if(su[i] == 1) ans.Add(i);
        }

        return ans;
    }
}

File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.cpp
class Solution {
public:
    void _eventualSafeNodes(vector<vector<int>>& graph,
                            vector<int>& vis,
                            vector<int>& su,
                            int startNode)
    {
        stack<int> stk;
        stk.push(startNode);

        while (!stk.empty()) {
            int currNode = stk.top();

            vis[currNode] = 1;
            bool allSafe = true;
            for (int neighbour : graph[currNode]) {
            if (vis[neighbour] == 0) {
                stk.push(neighbour);
                allSafe = false;
            } else if (su[neighbour] == 0 || su[neighbour] == -1) {
                allSafe = false;
                while(!stk.empty() && stk.top() != currNode){
                    stk.pop();
                }
                su[currNode] = 0;
                stk.pop();
                break;
            }
        }

            if (allSafe) {
                su[currNode] = 1;
                stk.pop();
            }
        }
    }
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        vector<int> vis(graph.size(), 0);
        vector<int> su(graph.size(), -1);
        
        int n = graph.size()-1;
        for(int j = n; j >= 0; j--)
        {
            if(vis[j] == 1) continue;
            _eventualSafeNodes(graph, vis, su, j);
        }

        vector<int> ans;
        for(int i=0;i<su.size();i++){
            if(su[i]) ans.push_back(i);
        }

        return ans;
    }
};